File: ui/client/src/lib/categoryResolver.ts
import categoriesJson from "@/data/categories.json";

export interface CategoryRule {
  duration_lt?: number;
  duration_gte?: number;
  weekday?: string;  // "Mon", "Tue", etc.
}

export interface CategoryEntry {
  code: string;
  label: string;
  rule?: CategoryRule;
}

export interface SportConfig {
  sport: string;
  categories: CategoryEntry[];
}

export type CategoryConfigObject = Record<string, Record<string, { label: string; rule?: CategoryRule }>>;
export type CategoryConfigInput = SportConfig[] | CategoryConfigObject;

export function resolveCategory(
  sportType: string,
  elapsedTime: number,
  startDateLocal: string,
  config: CategoryConfigInput
): string {
  const d = new Date(startDateLocal.replace(/Z$/, ""));
  const days = ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"];
  const weekday = days[d.getDay()];

  if (Array.isArray(config)) {
    const sportConfig = config.find(c => c.sport === sportType);
    if (!sportConfig) {
      return sportType.substring(0, 3).toUpperCase();
    }
    
    for (const entry of sportConfig.categories) {
      if (!entry.rule) return entry.code;
      
      let pass = true;
      if (entry.rule.duration_lt !== undefined && elapsedTime >= entry.rule.duration_lt) {
        pass = false;
      }
      if (entry.rule.duration_gte !== undefined && elapsedTime < entry.rule.duration_gte) {
        pass = false;
      }
      if (entry.rule.weekday !== undefined && entry.rule.weekday !== weekday) {
        pass = false;
      }
      
      if (pass) return entry.code;
    }
    
    return sportConfig.categories[0].code;
  }

  const sportMap = config[sportType];
  if (!sportMap || Object.keys(sportMap).length === 0) {
    return sportType.substring(0, 3).toUpperCase();
  }

  const codes = Object.keys(sportMap);
  for (const [code, info] of Object.entries(sportMap)) {
    if (!info.rule) return code;
    let pass = true;
    if (info.rule.duration_lt !== undefined && elapsedTime >= info.rule.duration_lt) {
      pass = false;
    }
    if (info.rule.duration_gte !== undefined && elapsedTime < info.rule.duration_gte) {
      pass = false;
    }
    if (info.rule.weekday !== undefined && info.rule.weekday !== weekday) {
      pass = false;
    }
    if (pass) return code;
  }

  return codes[0];
}

export function buildCategoryLookup(config: CategoryConfigInput): Record<string, { label: string; sport: string }> {
  const lookup: Record<string, { label: string; sport: string }> = {};

  if (Array.isArray(config)) {
    for (const sport of config) {
      for (const cat of sport.categories) {
        lookup[cat.code] = { label: cat.label, sport: sport.sport };
      }
    }
    return lookup;
  }

  for (const [sport, categories] of Object.entries(config)) {
    for (const [code, info] of Object.entries(categories)) {
      lookup[code] = { label: info.label, sport };
    }
  }

  return lookup;
}

let activeCategoryConfig: CategoryConfigInput | undefined = undefined;

export function setGlobalCategoryConfig(config?: CategoryConfigInput) {
  activeCategoryConfig = config;
}

export function getGlobalCategoryConfig(): CategoryConfigInput | undefined {
  return activeCategoryConfig;
}

const defaultCategoryLookup = buildCategoryLookup(categoriesJson as CategoryConfigInput);

/**
 * Derives the sport / discipline identifier (e.g. "badminton", "run", "calisthenics", "foundation", "cycling", "recovery", etc.)
 * for a given category code or category string using the category lookup config.
 */
export function getSportForCategory(category: string, config?: CategoryConfigInput): string {
  const cat = (category || "").trim();
  if (!cat) return "other";

  const effectiveConfig = config ?? activeCategoryConfig;
  const lookup = effectiveConfig ? buildCategoryLookup(effectiveConfig) : defaultCategoryLookup;
  
  // Direct code lookup (exact match or uppercase match)
  const entry = lookup[cat] || lookup[cat.toUpperCase()];
  if (entry) {
    const labelLower = (entry.label || "").toLowerCase();
    if (labelLower === "foundation") return "foundation";
    if (labelLower === "calisthenics") return "calisthenics";
    if (labelLower === "realign" || labelLower === "recovery") return "recovery";

    const sportLower = (entry.sport || "").toLowerCase();
    if (sportLower === "ride" || sportLower === "cycling" || sportLower === "bike") return "cycling";
    if (sportLower === "yoga") return "recovery";
    if (sportLower === "weighttraining" || sportLower === "weight_training" || sportLower === "strength") {
      return labelLower.includes("strength") ? "strength" : "weight_training";
    }
    if (["badminton", "run", "calisthenics", "foundation", "recovery", "hike", "walk", "cricket", "football", "workout", "swim"].includes(sportLower)) {
      return sportLower;
    }
  }

  // Fallback pattern / name matching for full category strings or direct sport names
  const lower = cat.toLowerCase();
  if (lower.startsWith("badminton")) return "badminton";
  if (lower === "calisthenics" || lower === "calisthenic") return "calisthenics";
  if (lower === "ride" || lower === "cycling" || lower === "bike") return "cycling";
  if (lower === "foundation") return "foundation";
  if (lower === "recovery" || lower === "realign" || lower === "mobility") return "recovery";
  if (lower === "run" || lower === "running") return "run";
  if (lower === "strength") return "strength";
  if (lower === "weight_training" || lower === "weights" || lower === "weighttraining") return "weight_training";
  if (lower === "hike") return "hike";
  if (lower === "walk") return "walk";
  if (lower === "cricket") return "cricket";
  if (lower === "football") return "football";
  if (lower === "workout") return "workout";
  if (lower === "swim") return "swim";

  return "other";
}

File: engine/core/category_resolver.py
import json
from datetime import datetime
from typing import Union, Dict, Any

def load_config(config_path: str) -> Union[Dict[str, Any], list]:
    """Load categories.json and return the config object."""
    with open(config_path, "r", encoding="utf-8") as f:
        return json.load(f)

def resolve_category(sport_type: str, elapsed_time: int, start_date_local: str, config: Union[Dict[str, Any], list]) -> str:
    """Resolve 3-letter category code from config.
    
    Supports both dictionary format:
      { "Badminton": { "RNK": { "label": "Ranked", "rule": ... } } }
    and list format:
      [ { "sport": "Badminton", "categories": [ { "code": "RNK", ... } ] } ]
    """
    if not sport_type:
        return "OTH"

    weekday_str = None
    if start_date_local:
        try:
            dt = datetime.fromisoformat(start_date_local.replace("Z", "+00:00"))
            weekdays = ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"]
            weekday_str = weekdays[dt.weekday()]
        except ValueError:
            pass

    # Dictionary format
    if isinstance(config, dict):
        sport_map = config.get(sport_type)
        if not sport_map or not isinstance(sport_map, dict):
            return sport_type[:3].upper()

        codes = list(sport_map.keys())
        if not codes:
            return sport_type[:3].upper()

        for code, info in sport_map.items():
            rule = info.get("rule") if isinstance(info, dict) else None
            if not rule:
                return code

            conditions_met = True
            if "duration_lt" in rule and elapsed_time >= rule["duration_lt"]:
                conditions_met = False
            if "duration_gte" in rule and elapsed_time < rule["duration_gte"]:
                conditions_met = False
            if "weekday" in rule and weekday_str != rule["weekday"]:
                conditions_met = False

            if conditions_met:
                return code

        return codes[0]

    # Array format (legacy / fallback)
    if isinstance(config, list):
        sport_config = next((item for item in config if item.get("sport") == sport_type), None)
        if not sport_config:
            return sport_type[:3].upper()

        categories = sport_config.get("categories", [])
        if not categories:
            return sport_type[:3].upper()

        for entry in categories:
            rule = entry.get("rule")
            if not rule:
                return entry["code"]

            conditions_met = True
            if "duration_lt" in rule and elapsed_time >= rule["duration_lt"]:
                conditions_met = False
            if "duration_gte" in rule and elapsed_time < rule["duration_gte"]:
                conditions_met = False
            if "weekday" in rule and weekday_str != rule["weekday"]:
                conditions_met = False

            if conditions_met:
                return entry["code"]

        return categories[0]["code"]

    return sport_type[:3].upper()

def resolve_from_activity(activity: Dict[str, Any], config: Union[Dict[str, Any], list]) -> str:
    """Convenience: resolve category from an activity dict."""
    sport_type = activity.get("sport_type", activity.get("type", ""))
    elapsed_time = activity.get("elapsed_time", 0)
    start_date_local = activity.get("start_date_local", "")
    return resolve_category(sport_type, elapsed_time, start_date_local, config)
